src/core/dev/ring.cpp

```diff
@@ -134,6 +143,112 @@ void ring::put_tcp_segs(tcp_seg *seg)
     put_obj_list(g_tcp_seg_pool, m_tcp_seg_list, seg, m_tcp_seg_count, return_treshold);
 }
 
+// Assumed ec is not nullptr
+void ring::socketxtreme_put_ecs(ring_ec *ec)
+{
+    static const uint32_t return_treshold = 256 * 2U;
+
+    std::lock_guard<decltype(m_ec_lock)> lock(m_ec_lock);
+
+    put_obj_list(g_socketxtreme_ec_pool, m_socketxtreme_ec_list, ec, m_socketxtreme_ec_count,
+                 return_treshold);
+}
+
+void ring::socketxtreme_ec_sock_list_add(sockinfo *sock)
+{
+    sock->set_ec_ring_list_next(nullptr);
+    if (likely(m_socketxtreme.ec_sock_list_end)) {
+        m_socketxtreme.ec_sock_list_end->set_ec_ring_list_next(sock);
+        m_socketxtreme.ec_sock_list_end = sock;
+    } else {
+        m_socketxtreme.ec_sock_list_end = m_socketxtreme.ec_sock_list_start = sock;
+    }
+}
+
+xlio_socketxtreme_completion_t &ring::socketxtreme_start_ec_operation(sockinfo *sock,
+                                                                      bool always_new)
+{
+    m_socketxtreme.lock_ec_list.lock();
+    if (likely(!sock->get_last_ec())) {
+        socketxtreme_ec_sock_list_add(sock);
+        always_new = true;
+    }
+
+    if (always_new) {
+        sock->add_ec(socketxtreme_get_ecs(1U));
+    }
+
+    return sock->get_last_ec()->completion;
+}
+
+void ring::socketxtreme_end_ec_operation()
+{
+    m_socketxtreme.lock_ec_list.unlock();
+}
+
+bool ring::socketxtreme_ec_pop_completion(xlio_socketxtreme_completion_t *completion)
+{
+    struct ring_ec *ec = nullptr;
+
+    m_socketxtreme.lock_ec_list.lock();
+    if (m_socketxtreme.ec_sock_list_start) {
+        ec = m_socketxtreme.ec_sock_list_start->pop_next_ec();
+
+        ring_logdbg(
```

Comment:
This is a frequent operation in data path, right? If yes - switch to "func" log level instead of debug.

Reply:
Done